pkg/tfvars/baremetal/baremetal_test.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMastersSelectionByRole(t *testing.T) {
+
+	cases := []struct {
+		scenario string
+
+		numControlPlaneReplicas int64
+		libvirtURI              string
+		apiVIP                  string
+		imageCacheIP            string
+		bootstrapOSImage        string
+		externalBridge          string
+		externalMAC             string
+		provisioningBridge      string
+		provisioningMAC         string
+		platformHosts           []*baremetal.Host
+		image                   string
+		ironicUsername          string
+		ironicPassword          string
+		ignition                string
+
+		expectedError string
+		expectedHosts []string
+	}{
+		{
+			scenario:                "filter-workers",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", "master"),
+				host("master-1", "master"),
+				host("worker-0", "worker"),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "mixed-order",
+			numControlPlaneReplicas: 1,
+			platformHosts: platformHosts(
+				host("worker-0", "worker"),
+				host("master-0", "master"),
+				host("worker-1", "worker"),
+			),
+			expectedHosts: []string{"master-0"},
+		},
+		{
+			scenario:                "hosts-norole",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", ""),
+				host("worker-0", "worker"),
+				host("master-1", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "no-role",
+			numControlPlaneReplicas: 3,
+			platformHosts: platformHosts(
+				host("master-0", ""),
+				host("master-1", ""),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1", "master-2"},
+		},
+		{
+			scenario:                "more-hosts-than-required",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", "master"),
+				host("master-1", "master"),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "more-hosts-than-required-norole",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("master-0", ""),
+				host("master-1", ""),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "more-hosts-than-required-mixed",
+			numControlPlaneReplicas: 2,
+			platformHosts: platformHosts(
+				host("worker-0", "worker"),
+				host("master-0", ""),
+				host("worker-1", "worker"),
+				host("master-1", ""),
+				host("worker-2", "worker"),
+				host("master-2", ""),
+			),
+			expectedHosts: []string{"master-0", "master-1"},
+		},
+		{
+			scenario:                "more-masters-than-required",
+			numControlPlaneReplicas: 1,
+			platformHosts: platformHosts(
+				host("server-0", ""),
+				host("server-1", ""),
+				host("server-2", "master"),
+			),
+			expectedHosts: []string{"server-0"},
```

Comment:
This doesn't seem right to me, because we are ignoring information that the user has explicitly specified and doing something else.

You could easily fix this by doing:

```
sortIndex := map[string]int{"master" : -1, "": 0, "worker": 1}
sort.SliceStable(platformHosts, func (i, j int) bool {return sortIndex[platformHosts[i].Role] < sortIndex[platformHosts[j].Role]}
```

and then just taking the first _n_ from the list as before.

Reply:
The original point was to try avoiding any behavioral change (for already working configurations), so that we could backport the fix in 4.8. In any case the lack of role specification is problematic, and IMO the best solution would be making the role tag as mandatory (which could be something to be introduced later).

@hardys wdyt?

Reply:
@hardys this will mean considering the [mixed case](https://github.com/openshift/installer/pull/5205#discussion_r733852440) as invalid.
Since I guess we all agree to make the role tag as mandatory in future, and since in the current implementation a clear warning is displayed in the output for the mixed case, maybe we could keep the current approach which anyhow address the original ordering issue.

Given an install-config.yaml with 5 hosts configured (2 with master role and 1 with worker role), that's the output:
```
$ openshift-install --dir=/test/cluster create cluster
WARNING Host ostest-master-norole-2 hasn't any role configured 
WARNING Host ostest-worker-norole-1 hasn't any role configured 
INFO Consuming Install Config from target directory
...
```